File: archives/discrete_optimization/week2/knapsack/java/Solver.java
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.IOException;
import java.util.List;

public class Solver {
    public static int OPTIMALITY = 1;

    public static void solve(List<String> lines) throws IOException {
        // Parse the lines from standard input and fill items
        String[] firstLine = lines.get(0).split("\\s+");
        int n = Integer.parseInt(firstLine[0]);
        int K = Integer.parseInt(firstLine[1]);

        int[] values = new int[n];
        int[] weights = new int[n];
        for(int i = 1; i <= n; i++){
            String line = lines.get(i);
            String[] parts = line.split("\\s+");
            values[i-1] = Integer.parseInt(parts[0]);
            weights[i-1] = Integer.parseInt(parts[1]);
        }

        // Initialize variables
        long[][] matrix = new long[n+1][K+1];
        int item_i, value_i, weight_i;
        int[] is_taken = new int[n];
        for (int i = 1; i <= n; i++) {
            value_i = values[i-1];
            weight_i = weights[i-1];
            for (int j = 1; j <= K; j++) {
                // Unselect the i-th item if the weight is bigger than the specified capacity
                if (j < weight_i) matrix[i][j] = matrix[i-1][j];
                // Otherwise select the i-th item if its value makes the sum of values bigger
                else matrix[i][j] = Math.max(matrix[i-1][j], matrix[i-1][j-weight_i] + value_i);
            }
        }

        // Select the highest total value
        long value = matrix[n][K];

        // Determine which items achieve the highest total value
        int j = K;
        for (int i = n; i > 0; i--) {
            if (matrix[i][j] != matrix[i-1][j]) {
                is_taken[i-1] = 1;
                j -= weights[i-1];
            }
        }

        // Prepare the solution in the specified output format
        System.out.println(value + " " + OPTIMALITY);
        for (int i=  0; i < n; i++){
            System.out.print(is_taken[i] + " ");
        }
        System.out.println("");
    }

    public static void main(String[] args) {
        // Get the filepath from standard input
        String fileName;
        if (args[0].startsWith("-file=")) fileName = args[0].substring(6);
        else fileName = args[0];
        Path filePath = Paths.get(fileName);

        // Read the file content and solve the knapsack problem
        try {
            List<String> lines = Files.readAllLines(filePath, StandardCharsets.UTF_8);
            solve(lines);
        }
        catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: archives/discrete_optimization/week2/knapsack/python/solver.py
#!/usr/bin/env pytyhon

from collections import namedtuple
from pathlib import Path
import sys
from typing import List


optimality: int = 1
item: namedtuple = namedtuple("Item", ['index', 'value', 'weight'])


def solve(input_: str) -> str:
    # Parse the lines from standard input and fill items
    lines: List[str] = input_.split('\n')
    n: int  # Number of items
    K: int  # Capacity
    n, K = map(int, lines[0].split(' '))

    items: List[namedtuple] = []
    for i in range(1, n+1):
        line = lines[i]
        parts = line.split(' ')
        items.append(item(i-1, int(parts[0]), int(parts[1])))

    # Initialize variables
    matrix: List[List[int]] = [[0 for _ in range(K+1)] for _ in range(n+1)]
    is_taken = [0 for _ in range(len(items))]
    item_i: int
    value_i: int
    weight_i: int
    for i in range(1, n+1):
        item_i = items[i-1]
        value_i = item_i.value
        weight_i = item_i.weight
        for j in range(1, K+1):
            # Unselect the i-th item if the weight is bigger than the specified capacity
            if j < weight_i:
                matrix[i][j] = matrix[i-1][j]
            # otherwise select the i-th item if its value makes the sum of values bigger
            else:
                matrix[i][j] = max(matrix[i-1][j], matrix[i-1][j-weight_i] + value_i)

    # Select the highest total value
    value: int = matrix[n][K]

    # Determine which items achieve the highest total value
    j: int = K
    for i in range(n, 0, -1):
        if matrix[i][j] != matrix[i-1][j]:
            is_taken[i-1] = 1
            j -= items[i-1].weight

    # Prepare the solution in the specified output format
    output: str = f"{value} {optimality}\n{' '.join(map(str, is_taken))}"
    return output


if __name__ == '__main__':
    if len(sys.argv) > 1:
        filepath: Path = Path(sys.argv[1].strip())
        with filepath.open() as f:
            input_ = f.read()
        output = solve(input_)
        print(output)
    else:
        print("This test requires an input file. Please select one from the data directory. (i.e. python solver.py ./data/ks_4_0)")
